From 364efc5f2c4104a3c9741394717288025199a7c0 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:11:31 -0400
Subject: [powerpc] powernv: Fix little endian issues in OPAL dump code

Message-id: <1410545655-205645-463-git-send-email-dzickus@redhat.com>
Patchwork-id: 94581
O-Subject: [RHEL7 PATCH 462/626] powerpc/powernv: Fix little endian issues in OPAL dump code
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 2d6b63bbdd7a649ea17fae19f8bab07d4b292a09
Author: Anton Blanchard <anton@samba.org>
Date:   Tue Apr 22 15:01:27 2014 +1000

    powerpc/powernv: Fix little endian issues in OPAL dump code

    Signed-off-by: Anton Blanchard <anton@samba.org>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/include/asm/opal.h b/arch/powerpc/include/asm/opal.h
index afb0fed..66ad7a7 100644
--- a/arch/powerpc/include/asm/opal.h
+++ b/arch/powerpc/include/asm/opal.h
@@ -868,8 +868,8 @@ int64_t opal_validate_flash(uint64_t buffer, uint32_t *size, uint32_t *result);
 int64_t opal_manage_flash(uint8_t op);
 int64_t opal_update_flash(uint64_t blk_list);
 int64_t opal_dump_init(uint8_t dump_type);
-int64_t opal_dump_info(uint32_t *dump_id, uint32_t *dump_size);
-int64_t opal_dump_info2(uint32_t *dump_id, uint32_t *dump_size, uint32_t *dump_type);
+int64_t opal_dump_info(__be32 *dump_id, __be32 *dump_size);
+int64_t opal_dump_info2(__be32 *dump_id, __be32 *dump_size, __be32 *dump_type);
 int64_t opal_dump_read(uint32_t dump_id, uint64_t buffer);
 int64_t opal_dump_ack(uint32_t dump_id);
 int64_t opal_dump_resend_notification(void);
diff --git a/arch/powerpc/platforms/powernv/opal-dump.c b/arch/powerpc/platforms/powernv/opal-dump.c
index 423f0f1..2fa59c8 100644
--- a/arch/powerpc/platforms/powernv/opal-dump.c
+++ b/arch/powerpc/platforms/powernv/opal-dump.c
@@ -214,15 +214,20 @@ static struct kobj_type dump_ktype = {
  .default_attrs = dump_default_attrs,
 };
 
-static int64_t dump_read_info(uint32_t *id, uint32_t *size, uint32_t *type)
+static int64_t dump_read_info(uint32_t *dump_id, uint32_t *dump_size, uint32_t *dump_type)
 {
+ __be32 id, size, type;
  int rc;
- *type = 0xffffffff;
 
- rc = opal_dump_info2(id, size, type);
+ type = cpu_to_be32(0xffffffff);
 
+ rc = opal_dump_info2(&id, &size, &type);
  if (rc == OPAL_PARAMETER)
-  rc = opal_dump_info(id, size);
+  rc = opal_dump_info(&id, &size);
+
+ *dump_id = be32_to_cpu(id);
+ *dump_size = be32_to_cpu(size);
+ *dump_type = be32_to_cpu(type);
 
  if (rc)
   pr_warn("%s: Failed to get dump info (%d)\n",
-- 
1.7.1